### PR TITLE
Improve User Code Deployment part

### PR DESCRIPTION
**Pull request by viliam-durina**
I recommend to review commit by commit.
@jerrinot Please check the part for member behavior
@sancar  Please check the client part

I did only superficially test what I wrote, the content is mostly based on discussions with other people.

### CODE REVIEW

File: src/docs/asciidoc/setting_up_clusters.adoc [hunk 1]
```diff
@@ -583,32 +580,39 @@ include::{javasource}/ClientUCD.java[tag=clientucd]
...
-the `LOCAL_AND_CACHED_CLASSES` mode, regardless of the configuration set on the
-member side (which can be `LOCAL_AND_CACHED_CLASSES`, `LOCAL_CLASSES_ONLY` and `OFF`).
+===== Two Versions of a Class
 
-The remaining properties, which are `blacklist-prefixes`, `whitelist-prefixes`
-and `provider-filter` configured on the member side, effect the client user code
-deployment's behavior too. For example, assuming that you provide `com.foo` as a
-blacklist prefix on the member side, the member discards the classes with the prefix
-`com.foo` loaded by the client.
+If the clients uploads a class and the member already has that class, an
```
Comment (Serdaro) [inline src/docs/asciidoc/setting_up_clusters.adoc]: If the client uploads ...

File: src/docs/asciidoc/setting_up_clusters.adoc [hunk 2]
```diff
@@ -583,32 +580,46 @@ include::{javasource}/ClientUCD.java[tag=clientucd]
...
+correctly:
+
+* User Code Deployment must be enabled on the members. Otherwise, the classes
+from the client will be ignored. Also blacklisted and non-whitelisted classes
+will be ignored.
+* All members must be providers, `provider-mode` must be set to
+`LOCAL_AND_CACHED_CLASSES` on all members.
+* No `provider-filter` must be configured.
+
+The client uploads the classes only to one member. If the members don't load
```
Comment (sancar) [inline src/docs/asciidoc/setting_up_clusters.adoc]: @viliam-durina I have checked the code. There is a small difference than what I described to you earlier. 
Client uploads the classes to one member, and classes send the other members over this single member in the beginning.  
But those classes are not migrated to new joining members. The new joining members relies on the member to member user code deployment.

File: src/docs/asciidoc/setting_up_clusters.adoc [hunk 3]
```diff
@@ -583,32 +580,46 @@ include::{javasource}/ClientUCD.java[tag=clientucd]
...
+
+Classes deployed from clients are always cached on the members, no matter
+whether `ETERNAL` or `OFF` is configured on the members.
+
+===== Performance Considerations
 
-For the property `class-cache-mode`, Client User Code Deployment supports only the
-`ETERNAL` mode, regardless of the configuration set on the member side
-(which can be `ETERNAL` and `OFF`).
+The member always uploads all added classes and jars to one of the members,
```
Comment (sancar) [inline src/docs/asciidoc/setting_up_clusters.adoc]: The client always uploads ....

File: src/docs/asciidoc/setting_up_clusters.adoc (hunk 2 above)
Comment (viliam-durina) [reply src/docs/asciidoc/setting_up_clusters.adoc]: my mind has parsing error here: :)
>and classes send the other members over this single member in the beginning

Is it that all current members somehow get the classes from the client when the client connects, only the newly joined members rely on member-to-member deployment? In that case a class-repository member will work...

Feel free to push to this PR with updated text. I'll then review and let you know if I understand it as a non-insider.

File: src/docs/asciidoc/setting_up_clusters.adoc [hunk 4]
```diff
@@ -393,35 +393,38 @@ during the cluster join process. Starting with 3.11, members no longer
...
+=== Deploying User Code on the Member
+
+Hazelcast can dynamically load your custom classes or domain classes from other
+members. A <<enabling-lite-members, lite member>> can be designated as a _class
+repository_, but any member can _provide_ classes to other members. For this
+purpose Hazelcast offers a distributed dynamic class loader.
+
+The following is a brief working mechanism of the User Code Deployment feature:
+
+. A new dynamic class loader is created to handle each operation.
```
Comment (jerrinot) [inline src/docs/asciidoc/setting_up_clusters.adoc]: what is "each operation" ?

Comment (viliam-durina) [reply src/docs/asciidoc/setting_up_clusters.adoc]: I summarized my understanding from a discussion, I couldn't actually find it in code:

image

**Pull request merged**